src/main/java/org/opentripplanner/routing/algorithm/raptor/transit/request/PatternGuaranteedTransferProvider.java

```diff
@@ -0,0 +1,177 @@
+        return currentTransfers != null;
+    }
+
+    @Override
+    public final RaptorTripScheduleBoardOrAlightEvent<TripSchedule> find(
+            RaptorTimeTable<TripSchedule> timetable,
+            TripSchedule sourceTripSchedule,
+            int sourceStopIndex,
+            int sourceArrivalTime
+    ) {
+        final Trip sourceTrip = sourceTripSchedule.getOriginalTripTimes().trip;
+        final int sourceStopPos = translator.findSourceStopPosition(
+                sourceTripSchedule, sourceArrivalTime, sourceStopIndex
+        );
+
+        TransferPoint targetPoint = findMatchingTargetPoint(sourceTrip, sourceStopPos);
+
+        if(targetPoint == null) { return null; }
+
+        return translator.findTimetableTripIndex(
+                timetable,
+                targetPoint.getTrip(),
+                targetPoint.getStopPosition(),
+                sourceArrivalTime
+        );
+    }
+
+    private TransferPoint findMatchingTargetPoint(
+            Trip sourceTrip,
+            int sourceStopPos
+    ) {
+    for (Transfer tx : currentTransfers) {
+            var sourcePoint = translator.source(tx);
+            if(sourcePoint.matches(sourceTrip, sourceStopPos)) {
+                return translator.target(tx);
+            }
+        }
+        return null;
+    }
+
+    private interface DirectionHelper {
+        TransferPoint source(Transfer tx);
+        TransferPoint target(Transfer tx);
+        int findSourceStopPosition(RaptorTripSchedule schedule, int timeLimit, int stop);
+        Result findTimetableTripIndex(
+                RaptorTimeTable<TripSchedule> timetable, Trip trip, int stopPos, int sourceTime
+        );
+    }
+
+    private static class ForwardDirectionHelper implements DirectionHelper {
+        @Override public TransferPoint source(Transfer tx) { return tx.getFrom();  }
+        @Override public TransferPoint target(Transfer tx) { return tx.getTo(); }
+        @Override
+        public int findSourceStopPosition(RaptorTripSchedule schedule, int timeLimit, int stop) {
+            return schedule.findArrivalStopPosition(timeLimit, stop);
+        }
+        @Override
+        public Result findTimetableTripIndex(
+                RaptorTimeTable<TripSchedule> timetable,
+                Trip trip,
+                int stopPos,
+                int sourceArrivalTime
+        ) {
+            // Abort after 6 hours
+            int maxLimit = sourceArrivalTime + 3600 * 6;
+
+            for (int i = 0; i < timetable.numberOfTripSchedules(); i++) {
+                var it = timetable.getTripSchedule(i);
+                int departureTime = it.departure(stopPos);
+                if(departureTime < sourceArrivalTime) { continue; }
+                if(departureTime > maxLimit) { return null; }
+                if(it.getOriginalTripTimes().trip == trip) {
+                    return new Result(i, it, stopPos, departureTime);
+                }
+            }
+            return null;
+        }
+    }
```

Comment:
This logic is "almost" the same as the logic in the implementation of `TripScheduleSearch`, but instead of refactoring this, I hope want to solve this by restructuring the transit model.  We duplicate 20 lines of code here, but it is not worth it to first refactor the `TripScheduleSearch` and then revert it when we get to cleaning up the transit model.